### DISCUSSION TITLE
Convergence impact of volumetric locking correction

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#21730] [Q&A Modules: Solid mechanics] [answered]
Hello,

I already opened a similar discussion once but I'm still having trouble with the volumetric locking correction in my frictional contact problems. The input I'm using is:

```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  volumetric_locking_correction = true
[]

[Mesh]
  [rad]
    type = AnnularMeshGenerator
    nt = 16
    nr = 12
    rmin = 17.5
    rmax = 40
    dmin = 225
    dmax = 315
  []
  [extrude]
    type = MeshExtruderGenerator
    input = rad
    extrusion_vector = '0 0 18'
    num_layers = 7
  []
  [transform]
    type = TransformGenerator
    input = extrude
    transform = translate
    vector_value = '0 40 0'
  []
  [rad_id]
    type = SubdomainIDGenerator
    input = transform
    subdomain_id = 1
  []
  [boden]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 8
    ny = 1
    nz = 5
    xmin = -35
    xmax = 35
    ymin = -1
    ymax = 0
    zmin = -7
    zmax = 25
    boundary_id_offset = 100
  []
  [boden_id]
    type = SubdomainIDGenerator
    input = boden
    subdomain_id = 2
  []
  [collect_meshes]
    type = MeshCollectionGenerator
    inputs = 'rad_id boden_id'
  []  
  patch_update_strategy = iteration
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    strain = FINITE
    block = '1 2'
  []
[]

[Contact]
  [kontakt]
    primary = top
    secondary = rmax
    formulation = penalty
    penalty = 1e4 
    model = coulomb
    friction_coefficient = 0.6
    normal_smoothing_distance = 0.1
  []
[]

[Dampers]
  [contact_slip]
    type = ContactSlipDamper
    primary = top
    secondary = rmax
  []
[]

[Functions]
  [load_disp]
    type = ParsedFunction
    value = '-3.7 * t'
  []
[]

[BCs]
  [load_disp]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = rmin
    function = load_disp
  []
  [ground_fixed_x]
    type = DirichletBC
    variable = disp_x
    boundary = bottom
    value = 0
  []
  [ground_fixed_y]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  [ground_fixed_z]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0
  []
  [wheel_fixed_z] 
    type = DirichletBC
    variable = disp_z
    boundary = rmin
    value = 0
  []
  [wheel_fixed_x] 
    type = DirichletBC
    variable = disp_x
    boundary = rmin
    value = 0
  []
[]

[Materials]
  [rad_elasticity]
    type = ComputeIsotropicElasticityTensor
    bulk_modulus = 320
    shear_modulus = 0.5
    block = '1'
  []
  [rad_stress]
    type = ComputeFiniteStrainElasticStress
    block = '1'
  []    
  [boden_elasticity]
    type = ComputeIsotropicElasticityTensor
    block = '2'
    youngs_modulus = 1e3
    poissons_ratio = 0.35
  []
  [boden_stress]
    type = ComputeFiniteStrainElasticStress
    block = '2'
  []
[]

[Preconditioning]
  [SMP]
    solve_type = 'PJFNK'
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_type'
    petsc_options_value = 'lu       superlu_dist'
  []
[]


[Executioner]
   type = Transient
   start_time = 0.0
   dt = 0.0125
   end_time = 1.0
   nl_abs_tol = 1e-6
   nl_rel_tol = 1e-6
   line_search = none
   [Predictor]
     type = SimplePredictor
     scale = 1.0
   []
    l_tol = 1e-3
   nl_max_its = 50
 []

[Outputs]
  exodus = true
  csv = true
[]

[Postprocessors]
  [nonlin_it]
    type = NumNonlinearIterations
  []
  [cumulative_nonlin_it]
    type = CumulativeValuePostprocessor
    postprocessor = nonlin_it
  []
  [lin_it]
    type = NumLinearIterations
  []
  [cumulative_lin_it]
    type = CumulativeValuePostprocessor
    postprocessor = lin_it
  []
  [time_sum]
    type = PerfGraphData
    section_name = "Root"
    data_type = total
    execute_on = 'initial nonlinear timestep_end'
  []
[]
```

It is meant to represent a segment of a rubber wheel getting pressed onto the ground. When not using locking correction there is locking which leads to incorrect contact behavior to the ground (the material in the contact zone is forced to expand to side in z direction instead of a more frictional sticky behaviour). I ran the simulation with two different shear moduli mu: 5 MPa and 0.5 MPa. The locking is worse for 0.5 MPa since the material is closer to incompressibility there. Then I ran both these simulations again but with locking correction activated.

What I observed regarding convergence:
- 0.5 MPa: 189 nonlinear iterations (nl it), 4182 linear iterations (l it)
- 5 MPa: 222 nl it, 10775 l it
- 0.5 MPa with locking correction: convergence so bad I aborted the simulation, would've taken too long
- 5 MPa with locking correction: 199 nl it, 6297 l it

What confuses me is the impact of the locking correction. Since there shouldn't be any locking regardless of the used shear modulus I'd expect the simulations that use the locking correction to behave similarily. But when mu = 0.5 MPa the locking correction has a harsh negative impact on convergence while with mu = 5 MPa it actually benefits convergence. 

Is there any known impact on convergence for locking correction? Or does anyone have an idea what causes this behaviour?

**Comment by lindsayad** [comment]
@recuero @jiangwen84 @bwspenc any suggestions for this user?

**Comment by recuero** [comment, marked as answer]
It may be that the Jacobian, when volumetric locking is used, is not as accurate as without locking correction. Any numerical impediment when using frictional contact will tend to hinder convergence noticeably. I agree that, in many problems, locking correction can be enabled and its practical impact on the results still be negligible, but, with contact, I wonder if the additional modes of deformation enabled by locking correction make the contact interface problem more challenging (i.e. can the combination of frictional and internal forces excite volumetric modes?).
Below an input file that uses automatic differentiation with your problem. I halved the time step. It solves in 4/5 nonlinear iterations and the number of linear iterations per nonlinear iteration is no more than 12/13.

That said, I do not fully understand these statements: 
> Since there shouldn't be any locking regardless of the used shear modulus

and 

> When not using locking correction there is locking which leads to incorrect contact behavior to the ground (the material in the contact zone is forced to expand to side in z direction instead of a more frictional sticky behaviour)

Do you expect locking correction to be relevant for the problem above? Are you using a different material model for your own testing?



<img width="2171" alt="image" src="https://user-images.githubusercontent.com/14300313/184412188-8022cb18-c820-427a-8c9b-870111bb06e5.png">

```[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  volumetric_locking_correction = true
[]

[Mesh]
  [rad]
    type = AnnularMeshGenerator
    nt = 16
    nr = 12
    rmin = 17.5
    rmax = 40
    dmin = 225
    dmax = 315
  []
  [extrude]
    type = MeshExtruderGenerator
    input = rad
    extrusion_vector = '0 0 18'
    num_layers = 7
  []
  [transform]
    type = TransformGenerator
    input = extrude
    transform = translate
    vector_value = '0 40 0'
  []
  [rad_id]
    type = SubdomainIDGenerator
    input = transform
    subdomain_id = 1
  []
  [boden]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 8
    ny = 1
    nz = 5
    xmin = -35
    xmax = 35
    ymin = -1
    ymax = 0
    zmin = -7
    zmax = 25
    boundary_id_offset = 100
  []
  [boden_id]
    type = SubdomainIDGenerator
    input = boden
    subdomain_id = 2
  []
  [collect_meshes]
    type = MeshCollectionGenerator
    inputs = 'rad_id boden_id'
  []
  patch_update_strategy = iteration
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    strain = FINITE
    block = '1 2'
    use_automatic_differentiation = true
  []
[]

[Contact]
  [kontakt]
    primary = top
    secondary = rmax
    formulation = penalty
    penalty = 1e1
    normalize_penalty = true
    model = coulomb
    friction_coefficient = 0.6
    normal_smoothing_distance = 0.1
  []
[]

[Dampers]
  [contact_slip]
    type = ContactSlipDamper
    primary = top
    secondary = rmax
  []
[]

[Functions]
  [load_disp]
    type = ParsedFunction
    value = '-3.7 * t'
  []
[]

[BCs]
  [load_disp]
    type = ADFunctionDirichletBC
    variable = disp_y
    boundary = rmin
    function = load_disp
  []
  [ground_fixed_x]
    type = ADDirichletBC
    variable = disp_x
    boundary = bottom
    value = 0
  []
  [ground_fixed_y]
    type = ADDirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  [ground_fixed_z]
    type = ADDirichletBC
    variable = disp_z
    boundary = bottom
    value = 0
  []
  [wheel_fixed_z]
    type = ADDirichletBC
    variable = disp_z
    boundary = rmin
    value = 0
  []
  [wheel_fixed_x]
    type = ADDirichletBC
    variable = disp_x
    boundary = rmin
    value = 0
  []
[]

[Materials]
  [rad_elasticity]
    type = ADComputeIsotropicElasticityTensor
    bulk_modulus = 320
    shear_modulus = 0.5
    block = '1'
  []
  [rad_stress]
    type = ADComputeFiniteStrainElasticStress
    block = '1'
  []
  [boden_elasticity]
    type = ADComputeIsotropicElasticityTensor
    block = '2'
    youngs_modulus = 1e3
    poissons_ratio = 0.35
  []
  [boden_stress]
    type = ADComputeFiniteStrainElasticStress
    block = '2'
  []
[]

[Preconditioning]
  [SMP]
    solve_type = 'PJFNK'
    type = SMP
    full = true

  []
[]

[Executioner]
  type = Transient
  start_time = 0.0
  dt = 0.00625
  end_time = 1.0
  nl_abs_tol = 1e-6
  nl_rel_tol = 1e-6
  line_search = none

  petsc_options_iname = '-pc_type -pc_factor_mat_solver_type'
  petsc_options_value = 'lu       superlu_dist'
  #[Predictor]
  #  type = SimplePredictor
  #  scale = 1.0
  #[]
  l_tol = 1e-3
  nl_max_its = 50
[]

[Outputs]
  exodus = true
  csv = true
[]

[Postprocessors]
  [nonlin_it]
    type = NumNonlinearIterations
  []
  [cumulative_nonlin_it]
    type = CumulativeValuePostprocessor
    postprocessor = nonlin_it
  []
  [lin_it]
    type = NumLinearIterations
  []
  [cumulative_lin_it]
    type = CumulativeValuePostprocessor
    postprocessor = lin_it
  []
  [time_sum]
    type = PerfGraphData
    section_name = "Root"
    data_type = total
    execute_on = 'initial nonlinear timestep_end'
  []
[]
```

**Reply by Flolaffel** [reply]
> I wonder if the additional modes of deformation enabled by locking correction make the contact interface problem more challenging (i.e. can the combination of frictional and internal forces excite volumetric modes?).

Sadly, I'm not able to answer this question. But assuming it was true that the enabled deformation modes make the problem more challenging, then why would locking correction improve convergence for mu = 5 and hinder it for mu = 0.5? 

> Below an input file that uses automatic differentiation with your problem. I halved the time step. It solves in 4/5 nonlinear iterations and the number of linear iterations per nonlinear iteration is no more than 12/13.

Is the picture you attached from said input? Because while you're right that the linear convergence improves massively with AD, running the input still results in a lot of failed time steps for me (negative Jacobian entries).

Is there any specific reason you commented out the predictor? Because in my experience it's quite a helpful tool for contact convergence in MOOSE which is also true for your provided input (it runs seemingly without failed time steps with predictor enabled).

> Do you expect locking correction to be relevant for the problem above? Are you using a different material model for your own testing?

Yes, volumetric locking seems to be a significant problem in the presented simulation. I just worded myself poorly. What I intented to say, was:

> Since there shouldn't be any locking regardless of the used shear modulus **when using locking correction**

So the quintessence of my statement was: I would expect the simulations that use locking correction to behave similarly regarding convergence impact, regardless of the shear modulus since volumetric locking should be non existent anymore. Problem is: they don't which is part of my confusion.

Furthermore, I used different material models for testing. The plan is to use UMAT as the final material model. There, the behavioral problems are similar to the built-in material model which is why I provided that one for the forum discussion.

Unluckily though your AD solution isn't compatible with the MOOSE UMAT class AFAIK.

**Reply by recuero** [reply]
> why would locking correction improve convergence for mu = 5 and hinder it for mu = 0.5?

I'd speculate, given the AD results, we can improve the Jacobian in that case. For mu = 0.5, Poisson's ratio is closer to incompressibility so the action of volumetric locking correction would be more relevant.

> Is the picture you attached from said input? Because while you're right that the linear convergence improves massively with AD, running the input still results in a lot of failed time steps for me (negative Jacobian entries).

> Is there any specific reason you commented out the predictor? Because in my experience it's quite a helpful tool for contact convergence in MOOSE which is also true for your provided input (it runs seemingly without failed time steps with predictor enabled).

Yeah, the picture is from that AD input file, which runs until the end of your simulation. There are a number of failed time steps in the simulation, especially as deformation becomes larger. Honestly, I didn't look too much into it, but it seemed to me that it was mostly cutting the time step once, or twice at the very end. You can certainly add the predictor for this problem, and I'd think it would help, just like you said. The only reason I removed the predictor was to assess convergence behavior without it.

> regardless of the shear modulus since volumetric locking should be non existent anymore

True. And AD seems to show that the Jacobian can be improved, because the more we need the locking correction, the more its Jacobian contribution will be relevant. 

> Unluckily though your AD solution isn't compatible with the MOOSE UMAT class AFAIK.

As things stand right now, you are right. We could potentially support a version of user subroutines that leverages AD objects for C++. Alternatively, you could take a stab at implementing your models within MOOSE's design, but this latter option is less appealing to you as far as I understand.

**Reply by Flolaffel** [reply]
> I'd speculate, given the AD results, we can improve the Jacobian in that case. For mu = 0.5, Poisson's ratio is closer to incompressibility so the action of volumetric locking correction would be more relevant.

Well then I guess I'll just have to deal with it for now. Thanks for opening the issue though.

> Alternatively, you could take a stab at implementing your models within MOOSE's design, but this latter option is less appealing to you as far as I understand.

I already have an AD version of the UMAT implemented (Neo Hooke). But you are right that using this implementation is less appealing than the UMAT right now since using the UMAT serves as a proof of concept for possible future projects with more complex UMATs. I guess I'll use both for now for comparison.

**Reply by hugary1995** [reply]
The new kernels in tensor mechanics + NEML could give you exact Jacobian using UMAT without AD.

**Reply by Flolaffel** [reply]
What is NEML?

**Reply by hugary1995** [reply]
I take it back, you don't need NEML here (NEML is another material library). So we've been developing this new system in tensor mechanics that gives you exact Jacobian without using AD. But an Abaqus UMAT interface is still missing in the new system. I can add that interface for UMAT and you should be able to get exact Jacobian (with volumetric locking correction) without AD. I should be able to do it this month. That being said, I don't guarantee an exact Jacobian will solve your convergence issue with contact.

**Reply by hugary1995** [reply]
Okay, so the PR is up: https://github.com/idaholab/moose/pull/21871

You can checkout the PR and test to see if it improves the convergence with VLC. Mind that your DDSDDE in the UMAT has to be exact in order for the overall Jacobian to be exact.

You can also be patient and wait for this PR to be merged. It's up to you.

**Reply by Flolaffel** [reply]
Thanks a lot! I was on holiday, but I'll check it out in the following days.

**Comment by recuero** [comment]
Issue on @Flolaffel's investigation: #21842.
